### ISSUE TITLE
Cannot solve the correct generic type from the usages when the `ParamSpec` contains generic types.

### ISSUE THREAD

**Issue by cainmagi** [microsoft/pyright#5495]
### Describe the bug

A `ParamSpec` is usually used to forward the input arguments of a function signature to another one. However, if the function providing the `ParamSpec` contains generic types, the incompatibility caused by the generic types may not be correctly detected, thus leading to more abnormal behaviors.

### **To Reproduce and Codes**

In this example, I deliberately feed the inconsistent types of `val2` and `val3`. However, `pyright` does not mark `val2` and `val3` as `Union` of two different types, or raise errors about this inconsistency,
```python
from typing import Callable, TypeVar
from typing_extensions import ParamSpec, reveal_type


P = ParamSpec("P")
S = TypeVar("S")
T = TypeVar("T")


def func1(val1: T, val2: S, val3: S) -> T:
    ...


def func2(func: Callable[P, T], *args: P.args, **kwargs: P.kwargs) -> T:
    ...


# val2 should has the same type of val3, but here raises no issues.
# val1 is incorrectly annotated as `T@func1`; this may be a designated feature, not a bug.
# val2 is incorrectly annotated as `S@func1`; this may be a designated feature, not a bug.
# val3 is annotated as int
# Type check does not know that val2 and val3 have the same type.
reveal_type(func2(func1, val1=1, val2=3, val3="s"))  # Annotated as `T@func1` which is expected to be `int` here.
reveal_type(func1(val1=1, val2=3, val3="s"))  # Annotated as `int` which is as expectation.
```
Although `val3` does not match the type of `val2`, `pyright` does not raise errors here. I think it is reasonable if `val2` and `val3` are treated as `int | str`. However, `pyright` seems to not know that `val2` and `val3` are `int | str`.

Furthermore, this "incorrect" (maybe not actually incorrect) usage even causes the returned type of `func2` to be incorrectly solved as `T@func1`, which is expected to be `int`.

If the usage of `func2` is simple and consistent, the annotation of the returned value will be correct.
```python
reveal_type(func2(func1, val1=1, val2="3", val3="s"))  # Annotated as `int`, which is as expectation.
```

#### Use only one `TypeVar`

It seems that this issue is not caused by using multiple generic types. Even if I only use one type, the same issue still exists:

```python
from typing import Callable, TypeVar
from typing_extensions import ParamSpec, reveal_type


P = ParamSpec("P")
T = TypeVar("T")


def func1(val1: T, val2: T) -> T:
    ...


def func2(func: Callable[P, T], *args: P.args, **kwargs: P.kwargs) -> T:
    ...


# Both of the following two lines do not throw any errors in the basic mode.
reveal_type(func2(func1, val1=1, val2=2))  # Annotated as `int`, which is as expectation.
reveal_type(func2(func1, val1=1, val2="2"))  # Annotated as `T@func1`, should this be `int | str`?
reveal_type(func1(val1=1, val2="2"))  # Annotated as `int | str`, which is as expectation.
```

#### Use `ParamSpec` in `Protocol`

A similar issue will exist if `ParamSpec` is used in the `Protocol`

```python
from typing import Generic, Type, TypeVar
from typing_extensions import ParamSpec, Protocol, reveal_type


P = ParamSpec("P")
T = TypeVar("T")
T_co = TypeVar("T_co", covariant=True)


class TestProtocol(Protocol[P, T_co]):
    def __init__(self, val1: T_co, *args: P.args, **kwargs: P.kwargs) -> None:
        ...


def func2(cls: Type[TestProtocol[P, T]], val1: T, *args: P.args, **kwargs: P.kwargs) -> T:
    ...


class Test(Generic[T]):
    def __init__(self, val1: T, val2: int) -> None:
        ...


reveal_type(func2(Test, val1=1, val2=3))  # Annotated as `T@test | int`, however, expecting `int`.
```

### `mypy` behaves differently compared to `pyright`

I have used `mypy` to perform similar checks. To my surprise, it seems that `mypy` rejects all usages like this. However, I think it is quite common to pass a function with generic types as a callable argument providing the `ParamSpec`. So I think `pyright` should be better than `mypy` here.

```python
from typing import Callable, TypeVar
from typing_extensions import ParamSpec, reveal_type


P = ParamSpec("P")
S = TypeVar("S")
T = TypeVar("T")


def func1(val1: T, val2: S, val3: S) -> T:
    return val1


def func2(func: Callable[P, T], *args: P.args, **kwargs: P.kwargs) -> T:
    return func(*args, **kwargs)


reveal_type(func2(func1, val1=1, val2=2, val3=3))
```
It seems that `mypy` is also unable to detect the correct signature of `func2`:

```log
check_type.py:18: note: Revealed type is "T`-1"
check_type.py:18: error: Argument "val1" to "func2" has incompatible type "int"; expected "T"  [arg-type]
check_type.py:18: error: Argument "val2" to "func2" has incompatible type "int"; expected "S"  [arg-type]
check_type.py:18: error: Argument "val3" to "func2" has incompatible type "int"; expected "S"  [arg-type]
Found 3 errors in 1 file (checked 1 source file)
```

However, different form `pyright`, `mypy` is even unable to detect the type of input arguments in this example:
```python
from typing import Callable, TypeVar
from typing_extensions import ParamSpec, reveal_type


P = ParamSpec("P")
T = TypeVar("T")


def func1(val1: T) -> T:
    return val1


def func2(func: Callable[P, T], *args: P.args, **kwargs: P.kwargs) -> T:
    return func(*args, **kwargs)


reveal_type(func2(func1, val1=1))
```
Here is the error logs of `mypy`
```log
check_type.py:17: note: Revealed type is "T`-1"
check_type.py:17: error: Argument "val1" to "func2" has incompatible type "int"; expected "T"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

### **Expected behavior**

My expectation is that `func2` and `func1` in the following example should behave similarly when they are used with `func3`:

```python
from typing import Callable, TypeVar
from typing_extensions import ParamSpec, reveal_type


P = ParamSpec("P")
S = TypeVar("S")
T = TypeVar("T")


def func1(val1: T, val2: T) -> T:
    return val1


def func2(val1: T, val2: S, val3: S) -> T:
    ...


def func3(func: Callable[P, T], *args: P.args, **kwargs: P.kwargs) -> T:
    ...


reveal_type(func3(func1, val1=1, val2="3"))  # Annotated as `T@func1` which is expected to be `int | str` here or raise an inconsistency error of `val2` and `val3`.
reveal_type(func1(val1=1, val2="3"))  # Annotated as `int | str` which is as expectation. This result should be the same as the previous line.

reveal_type(func3(func2, val1=1, val2=2, val3="3"))  # Annotated as `T@func2` which is expected to be `int` here or raise an inconsistency error of `val2` and `val3`.
reveal_type(func2(val1=1, val2=2, val3="3"))  # Annotated as `int` which is as expectation. This result should be the same as the previous line.
```

### **VS Code extension or command-line**

* **Pylance** version: 2023.7.20
* **Pyright** version: 74d8f3c4

**Comment by erictraut** [comment]
Mypy doesn't have any support for higher-order functions that are passed generic functions, so it's not surprising that it doesn't handle this case. I've spent significant time to make this work in pyright. If you're interested in some examples, look at the test cases that begin with `solverHigherOrder` in the pyright source base.

You've identified a hole in pyright's current support for higher-order functions — in particular, where the higher-order function is parameterized by a ParamSpec. This will take some additional complex logic to handle. I don't consider it a high priority, so it may take me some time to get to this issue relative to some of the other open issues.

**Comment by cainmagi** [comment]
That sounds tricky! I remember that when I worked with a large project, the nested usages of `map(...)`, `filter(...)` and other iterators may also raise errors not supposed to happen. Refreshing the `pyright` can make such errors disappeared. I am not sure whether such cases are also caused by high-order type passing. Since I cannot reproduce those cases, I have not mentioned these phenomena before.

I will add more information if I get any more inspections related to this issue. Currently, I can simply avoid such usages until this issue is tackled. 

Thank you!

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.319, which I just published. It will also be included in a future release of pylance.
